docs/proposals/11-cluster-health.md

```diff
@@ -0,0 +1,298 @@
+
+### Gardener Shoot Resource
+
+The `Shoot` [resource](https://github.com/gardener/gardener/blob/master/pkg/apis/garden/types.go#L776) represents the desired and actual state of a shoot cluster.
+It has the `status.conditions` field which is updated by many components including the [shoot care controller](https://github.com/gardener/gardener/blob/master/pkg/gardenlet/controller/shoot/shoot_care_control.go) to [maintain](https://github.com/gardener/gardener/blob/master/pkg/gardenlet/controller/shoot/shoot_care_control.go#L160) shoot cluster health conditions.
+
+This is the standard API to publish such cluster health conditions to the end-users (administrators) for manual or automated consumption.
+
+### Gardener Extensions Cluster Resource
+
+The `Cluster` [resource](https://github.com/gardener/gardener/blob/master/pkg/apis/extensions/v1alpha1/types_cluster.go#L30) is part of the Gardener [extensions API](https://github.com/gardener/gardener/blob/master/pkg/apis/extensions/v1alpha1/) and is used as a point of co-ordination for the different extensions and controllers in the seed cluster to maintain spec, state and the co-ordinate action for each shoot cluster whose contol-plane is hosted in that seed cluster.
+
+This resource is [non-namespaced](https://github.com/gardener/gardener/blob/master/pkg/apis/extensions/v1alpha1/types_cluster.go#L26) and the part to maintain the shoot status conditions in it is [unstructured](https://github.com/gardener/gardener/blob/master/pkg/apis/extensions/v1alpha1/types_cluster.go#L57).
+
+## Proposal
+
+Extending the analogy of [node-problem-detector](#node-problem-detector) to the larger granularity of the Kubernetes cluster, the [`ClusterCondition`s](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L147) in the[`status`](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L66) of the [`Cluster`](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L30) custom resource (as defined in the [cluster-registry](#cluster-registry)) can be used as the standard way to publish and consume cluster health conditions.
+
+I.e.
+- A single `Cluster` resource instance is maintained for each cluster in the shoot control-plane namespace of the seed cluster.
```

Comment:
As far as I understand you plan to introduce a new `Cluster` resource (what a naming we have - we have Shoot, we have Cluster in `extensions.gardener.cloud` API group and we now want to introduce one more `Cluster`) and to maintain in the `.status.conditions` of this Cluster resource a new set of conditions.

Your proposed actions are in form of 
> If `KubeAPIServerReachableExternally ` is `false`, the `kube-controller-manager` could be scaled down to avoid marking `nodes` and `pods` as `NotReady` unnecessarily.

Which component will perform the `possible actions`?
Will `gardenlet`, `extension-controller` be responsible for watching/acting on this new Cluster resource?
If it is a new component that will perform the remedy actions, how we ensure that there won't be conflicts?

Could you please give some insights about how the conditions are aggregated in this new Cluster resource? How the possible actions are executed in terms of components which we currently have.

Reply:
> what a naming we have

Naming is hard and unforgiving :-)

> > If `KubeAPIServerReachableExternally ` is `false`, the `kube-controller-manager` could be scaled down to avoid marking `nodes` and `pods` as `NotReady` unnecessarily.
> 
> Which component will perform the `possible actions`?

To begin with, [this](https://github.com/gardener/gardener/pull/1797/files/0066744bd405637094f0df87e8671dce31b08002#diff-dc2b4cab766fb3ad06a4e681a236b570R192).

Reply:
> Will gardenlet, extension-controller be responsible for watching/acting on this new Cluster resource?

If then the gardenlet (the shoot care controller to be precise) can update the conditions in the cluster-registry `Cluster` resource (or in extensions.gardener.cloud/v1alpha1.Cluster` if we follow @rfranzke's [suggestion](https://github.com/gardener/gardener/pull/1797#discussion_r371294678)) and then copy whatever conditions that relevant to the `Shoot` resource `status.conditions`.

Then, any controller (extension or otherwise) can watch and take action if it has some meaningful action to take.

Reply:
> If it is a new component that will perform the remedy actions, how we ensure that there won't be conflicts?

The same we are doing right now. We as gardener developers and the landscape admins would have to know about such conflicts between different extensions and controllers before we introduce them into the landscape.

Reply:
> Could you please give some insights about how the conditions are aggregated in this new Cluster resource? How the possible actions are executed in terms of components which we currently have.

Explained below [here](https://github.com/gardener/gardener/pull/1797/files/0066744bd405637094f0df87e8671dce31b08002#diff-dc2b4cab766fb3ad06a4e681a236b570R177) and  [here](https://github.com/gardener/gardener/pull/1797/files/0066744bd405637094f0df87e8671dce31b08002#diff-dc2b4cab766fb3ad06a4e681a236b570R223). There might be other use-cases in the future.

Please let me know if the explanation below is not clear. I will try and elaborate.

Reply:
> Then, any controller (extension or otherwise) can watch and take action if it has some meaningful action to take.

Okay, could you please adapt the GEP with details which component watches for what and which component takes what kind of actions? I guess this is the info that I currently miss.

Reply:
> Okay, could you please adapt the GEP with details which component watches for what and which component takes what kind of actions? I guess this is the info that I currently miss.

Actually, I will remove the future use-cases as already discussed :-)

I will file the proposals/PRs as and when the future use-cases become priority.